common/environment/environment.cpp

```diff
@@ -1829,14 +1829,16 @@ unsigned CConstDropZoneInfoIterator::count() const
 
 static CriticalSection getEnvSect;
 
-extern ENVIRONMENT_API IEnvironmentFactory * getEnvironmentFactory()
+extern ENVIRONMENT_API IEnvironmentFactory * getEnvironmentFactory(bool update)
 {
     CriticalBlock block(getEnvSect);
     if (!factory)
     {
         factory = new CEnvironmentFactory();
         addShutdownHook(*factory);
     }
+    if (update)
```

Comment:
Is this safe?

Any previous callers may still be holding on to their copies of factory, I think

Reply:
Good question. Should I move those 2 lines after line 1837?

Reply:
No.

I don't think it's a new issue - but if I am reading this code correctly the code in environment.cpp is working on the assumption that it is managing a singleton that is never updated, but that assumption was rendered invalid once calls to validateCache were added.

@jakesmith What do you think?

Reply:
They should be called after line 1839. By reading the code inside validateCache(), there are 2 functions: create the subscription and run handleEnvironmentChange. The handleEnvironmentChange calls clearCache(). Inside the clearCache(), the synchronized procedure(safeCache) should handle the thread safe, I think.

Reply:
>it is managing a singleton that is never updated, but that assumption was rendered invalid once calls to validateCache were added.

There is only ever 1 factor (and 1 subscription).
The 1st call to validateCache will install the subscription. I though the idea then was that any notifications could clear the cache, so that subsequent calls to e.g. factory->openEnvironment() would reflect the update and not see old cached version.

However, handleEnvironmentChange() which resets the cache, is never automatically called. I would have expected it to have been called if the dirty flag had been set by the notifier on the next call to get the environment (e.g. next openEnvironment()).

Reply:
@richardkchapman - I think it's thread safe, _if_ every interaction with the environment (via IConstEnvironment), locks out the cache clearing , including anything that's returned like e.g. a IConstDropZoneInfoIterator). 
And by the looks of it, that is the intent, every call is protected inside a 'synchronized procedure(safeCache)' that is locked when the cache is cleared.

Reply:
I think it is probably safe as well.